Lib/inspect.py

```diff
@@ -60,6 +60,133 @@
+        authoritative source instead of the heuristic.
+
+    globals and locals are passed in to eval(); see the documentation
+    for eval() for more information.  If globals is None, this function
+    uses a context-specific default value, contingent on type(obj):
+
+      * If obj is a module, globals defaults to obj.__dict__.
+      * If obj is a class, globals defaults to
+        sys.modules[obj.__module__].__dict__.
+      * If obj is a callable, globals defaults to obj.__globals__,
+        although if obj is a wrapped function (using
+        functools.update_wrapper()) it is first unwrapped.
+
+    If get_annotations() doesn't have a usable globals dict--one
+    was not passed in, and the context-specific globals dict
+    was not set--and eval_str is inspect.ONLY_IF_STRINGIZED,
+    get_annotations() will behave as if eval_str is false.
+    """
+    if isinstance(obj, type):
+        # class
+        obj_dict = getattr(obj, '__dict__', None)
+        if obj_dict and hasattr(obj_dict, 'get'):
+            ann = obj_dict.get('__annotations__', None)
+        else:
+            ann = None
+
+        obj_globals = None
+        module_name = getattr(obj, '__module__', None)
+        if module_name:
+            module = sys.modules.get(module_name, None)
+            if module:
+                obj_globals = getattr(module, '__dict__', None)
+        unwrap = obj
+    elif isinstance(obj, types.ModuleType):
+        # module
+        ann = getattr(obj, '__annotations__', None)
+        obj_globals = getattr(obj, '__dict__')
+        unwrap = None
+    elif callable(obj):
+        # this includes types.Function, types.BuiltinFunctionType,
+        # types.BuiltinMethodType, functools.partial, functools.singledispatch,
+        # "class funclike" from Lib/test/test_inspect... on and on it goes.
+        ann = getattr(obj, '__annotations__', None)
+        obj_globals = getattr(obj, '__globals__', None)
+        unwrap = obj
+    else:
+        raise TypeError(f"{obj!r} is not a module, class, or callable.")
+
+    if ann is None:
+        return {}
+
+    if not isinstance(ann, dict):
+        raise ValueError(f"{obj!r}.__annotations__ is neither a dict nor None")
+
+    if not ann:
+        return {}
+
+    if eval_str is ONLY_IF_STRINGIZED:
+        eval_str = bool(obj_globals) and all(isinstance(v, str) for v in ann.values())
+    if not eval_str:
+        return dict(ann)
+
+    if unwrap is not None:
+        while True:
+            if hasattr(unwrap, '__wrapped__'):
+                unwrap = unwrap.__wrapped__
+                continue
+            if isinstance(unwrap, functools.partial):
+                unwrap = unwrap.func
+                continue
+            break
+        if hasattr(unwrap, "__globals__"):
+            obj_globals = unwrap.__globals__
+
+    if globals is None:
+        globals = obj_globals
+
+    return_value = {key:
+        value if not isinstance(value, str) else eval(value, globals, locals)
```

Comment:
Drive-by comment: in get_type_hints() we recently introduced a feature (promised by PEP 563 but not previously implemented) where the locals are given a correct default for class annotations (GH-24201). Maybe we should do that here too?

Reply:
Good point!  I'll take a look, once I get a PR for bpo-43901 out the door.